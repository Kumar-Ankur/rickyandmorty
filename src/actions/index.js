import * as actionTypes from "../types";

/**
 * @function getCharacters
 * @description action types for fetching characters data from API using saga file and update the state.
 */
export function getCharacters() {
  return { type: actionTypes.GET_CHARACTERS };
}

/**
 * @function getFilteredData
 * @param { Object } data - Filtered data
 * @description - filtered data based on payload value.
 */
export function getFilteredData(data) {
  return { type: actionTypes.GET_FILTERED_DATA, payload: data };
}

/**
 * @function sortAscendingById
 * @description - sort all the character data in ascending order
 */
export function sortAscendingById() {
  return { type: actionTypes.SORT_DATA_BY_ASCENDING };
}

/**
 * @function sortDescendingById
 * @description - sort all the character data in descending order.
 */
export function sortDescendingById() {
  return { type: actionTypes.SORT_DATA_BY_DESCENDING };
}

/**
 * @function addFilteredData
 * @param { String } data
 * @description - add filtered data in string for species category state
 */
export function addFilteredData(data) {
  return { type: actionTypes.ADD_FILTERED_DATA, payload: data };
}

/**
 * @function removeFilteredData
 * @param { String } data
 * @description - remove filtered data in string for species category state
 */
export function removeFilteredData(data) {
  return { type: actionTypes.REMOVE_FILTERED_DATA, payload: data };
}

/**
 * @function addGenderFilteredData
 * @param { String } data
 * @description - add filtered data in string for gender category state
 */
export function addGenderFilteredData(data) {
  return { type: actionTypes.ADD_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function removeGenderFilteredData
 * @param { String } data
 * @description - remove filtered data in string for gender category state
 */
export function removeGenderFilteredData(data) {
  return { type: actionTypes.REMOVE_GENDER_FILTERED_DATA, payload: data };
}

/**
 * @function addOriginFilteredData
 * @param { String } data
 * @description - add filtered data in string for origin category state
 */
export function addOriginFilteredData(data) {
  return { type: actionTypes.ADD_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function removeOriginFilteredData
 * @param { String } data
 * @description - remove filtered data in string for origin category state
 */
export function removeOriginFilteredData(data) {
  return { type: actionTypes.REMOVE_ORIGIN_FILTERED_DATA, payload: data };
}

/**
 * @function addSearchText
 * @param { String } data
 * @description - add search data in string for search category state
 */
export function addSearchText(data) {
  return { type: actionTypes.ADD_SEARCH_TEXT, payload: data };
}
