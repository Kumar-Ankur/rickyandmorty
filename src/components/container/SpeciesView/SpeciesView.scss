.species-content {
    color: white;
    border: 1px solid white;
    // display: inline-block;
    padding: 2rem;
    margin-left: 2rem;
}