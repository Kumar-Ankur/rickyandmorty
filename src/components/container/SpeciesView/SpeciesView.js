import React, { Component } from "react";
import CheckBox from "../../presentation/CheckBox/CheckBox";
import { connect } from "react-redux";
import * as actionCreator from "../../../actions";
import * as constants from "../../../constants";
import { getCheckBoxData } from '../../../utils';
import './SpeciesView.scss';

class SpeciesView extends Component {

  render() {
    const { all_characters } = this.props;
    const speciesData = getCheckBoxData(all_characters, constants.SPECIES);
    const { addFilteredData, removeFilteredData } = this.props;

    const getSpeciesBox = speciesData.map((data, index) => {
      return (
        <div key={index}>
          <CheckBox
            data={data}
            addFilter={addFilteredData}
            removeFilter={removeFilteredData}
          />
        </div>
      );
    });

    return (
      <div className="species-content">
          <h2>{ constants.SPECIES_FILTER }</h2>
        <form>{getSpeciesBox}</form>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    all_characters: state.character.all_characters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    addFilteredData: (data) => dispatch(actionCreator.addFilteredData(data)),
    removeFilteredData: (data) =>
      dispatch(actionCreator.removeFilteredData(data)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SpeciesView);
