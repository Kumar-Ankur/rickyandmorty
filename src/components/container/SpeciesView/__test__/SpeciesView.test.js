import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import SpeciesView, { mapDispatchToProps } from '../SpeciesView';
import { applyMiddleware, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import rootReducer from '../../../../reducers';
import * as actionCreator from "../../../../actions";

const middleware = createSagaMiddleware();
const mockStore = createStore(rootReducer, applyMiddleware(middleware));

function setUp(props={}) {
    return shallow(<SpeciesView store={ mockStore} {...props}/>)
}

function findByTestData(wrapper, val) {
    return wrapper.find(`[data-test='${val}']`);
}

describe('check species component', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = setUp().dive().dive();
    });

    test('generate snapshot for species component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('check species container render without crash', () => {
        const speciesComponent = findByTestData(wrapper, 'component-species');
        expect(speciesComponent.length).toBe(1);
    });

    test('check species form container render without error', () => {
        const formComponent = findByTestData(wrapper, 'component-species-form');
        expect(formComponent.length).toBe(1);
    });

    test('check `addFilteredData` working inside mapDispatchToProps', () => {
        const dispatchSpy = sinon.spy();
        mockStore.dispatch = dispatchSpy;
        const { addFilteredData } = mapDispatchToProps(dispatchSpy);
        addFilteredData('Human');
        const lastSpyCall = dispatchSpy.args[0][0];
        const expectedAction = actionCreator.addFilteredData('Human');
        expect(lastSpyCall.types).toEqual(expectedAction.types);
    });

    test('check `removeFilteredData` working inside mapDispatchToProps', () => {
        const dispatchSpy = sinon.spy();
        mockStore.dispatch = dispatchSpy;
        const { removeFilteredData } = mapDispatchToProps(dispatchSpy);
        removeFilteredData('Human');
        const lastSpyCall = dispatchSpy.args[0][0];
        const expectedAction = actionCreator.removeFilteredData('Human');
        expect(lastSpyCall.types).toEqual(expectedAction.types);
    });
})