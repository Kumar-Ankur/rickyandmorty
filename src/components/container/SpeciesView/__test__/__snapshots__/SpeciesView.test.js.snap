// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check species component generate snapshot for species component 1`] = `ShallowWrapper {}`;
